src/jshint.js

```diff
@@ -1284,10 +1284,11 @@ var JSHINT = (function() {
       "Array", "ArrayBuffer", "Boolean", "Collator", "DataView", "Date",
       "DateTimeFormat", "Error", "EvalError", "Float32Array", "Float64Array",
       "Function", "Infinity", "Intl", "Int16Array", "Int32Array", "Int8Array",
-      "Iterator", "Number", "NumberFormat", "Object", "RangeError",
-      "ReferenceError", "RegExp", "StopIteration", "String", "SyntaxError",
+      "Iterator", "Map", "Number", "NumberFormat", "Object", "Promise", "Proxy", "RangeError",
+      "Symbol", "Set",
+      "ReferenceError", "Reflect", "RegExp", "StopIteration", "String", "SyntaxError",
       "TypeError", "Uint16Array", "Uint32Array", "Uint8Array", "Uint8ClampedArray",
-      "URIError"
+      "URIError", "WeakMap", "WeakSet"
```

Comment:
How about this:

```js
var _natives = null;
var _nativesVer = NaN;
var baseNatives = [
      "Array", "ArrayBuffer", "Boolean", "Collator", "DataView", "Date",
      "DateTimeFormat", "Error", "EvalError",
      "Function", "Infinity", "Intl",
      "Number", "NumberFormat", "Object", "RangeError",
      "ReferenceError", "RegExp", "StopIteration", /* what even is this? Intl thing? */ "String", "SyntaxError",
      "TypeError",
      "URIError"
];
var ecma2015Natives = [
  "Iterator", "Float32Array", "Float64Array", "Int16Array", "Int32Array", "Int8Array",
  "Uint16Array", "Uint32Array", "Uint8Array", "Uint8ClampedArray",
  "WeakMap", "WeakSet", "Symbol", "Map", "Set", "Proxy", "Promise",
];

function getNativesNames() {
  if (_nativesVer === state.option.esversion) return _natives;
  
  _nativesVer = state.option.esversion;
  _natives = _.assign([], baseNatives);
  if (_nativesVer > 6) _.assign(_natives, ecma2015Natives);
  return _natives;
}
```

this gives allows inline `esversion` directives to be honoured within the program, and lets us avoid checking modern builtins in legacy environments